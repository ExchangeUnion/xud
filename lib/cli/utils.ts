import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
