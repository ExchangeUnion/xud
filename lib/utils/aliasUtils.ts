import { wordlist } from '../constants/wordlist';
import errors from '../service/errors';

/**
 * Gets a two-word alias from a node pubkey. Throws an error if the pubkey is
 * not a valid 66 character hex string.
 * @param pubkey a public key in hex format to convert to an alias
 */
export const pubKeyToAlias = (pubkey: string): string => {
  const getAlias = (substring: string): string | undefined => {
    const asNumber = parseInt(substring, 16);
    if (!isNaN(asNumber)) {
      const { length } = wordlist;
      const index = asNumber % (length * length);
      const alias = wordlist[Math.floor(index / length)] + wordlist[index % length];
      return alias;
    }
    return undefined;
  };
  if (pubkey.length === 66) {
    const alias = getAlias(pubkey);
    if (alias) {
      return alias;
    }
  }
  throw errors.INVALID_ARGUMENT('Not a valid public key');
};

/**
 * Returns true if a given node identifier could be a node pub key, namely
 * if it is a hex string with the correct length.
 */
export const isNodePubKey = (nodeIdentifier: string): boolean => {
  return nodeIdentifier.length === 66 && /^[0-9a-fA-F]+$/.test(nodeIdentifier);
};
