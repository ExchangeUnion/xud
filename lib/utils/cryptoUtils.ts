import { createCipheriv, createDecipheriv, createHash, randomBytes as cryptoRandomBytes } from 'crypto';
import { promisify } from 'util';

const ENCRYPTION_IV_LENGTH = 16;

/** A promisified wrapper for the NodeJS `crypto.randomBytes` method. */
export const randomBytes = promisify(cryptoRandomBytes);

function getCipherKey(password: string) {
  return createHash('sha256').update(password).digest();
}

/**
 * Encrypts a Buffer or base64 string using a password
 * @param payload a Buffer or base64 string
 * @returns an encrypted Buffer
 */
export async function encrypt(payload: Buffer | string, password: string) {
  const buf = typeof payload === 'string' ? Buffer.from(payload, 'utf8') : payload;

  const iv = await randomBytes(ENCRYPTION_IV_LENGTH);
  const key = getCipherKey(password);
  const cipher = createCipheriv('aes-256-cbc', key, iv);
  return Buffer.concat([iv, cipher.update(buf), cipher.final()]);
}

/**
 * Decrypts a Buffer or base64 string using a password
 * @param payload a Buffer or base64 string
 * @returns a decrypted Buffer
 */
export function decrypt(payload: Buffer | string, password: string) {
  const buf = typeof payload === 'string' ? Buffer.from(payload, 'base64') : payload;

  // the first 16 bytes contain the initialization vector
  const iv = buf.slice(0, ENCRYPTION_IV_LENGTH);
  const key = getCipherKey(password);
  const encrypted = buf.slice(ENCRYPTION_IV_LENGTH);
  const decipher = createDecipheriv('aes-256-cbc', key, iv);
  return Buffer.concat([decipher.update(encrypted), decipher.final()]);
}

/** Returns a random payment preimage and hash in hex encoding. */
export async function generatePreimageAndHash() {
  const bytes = await randomBytes(32);
  const rPreimage = bytes.toString('hex');
  const rHash = createHash('sha256').update(bytes).digest('hex');
  return { rPreimage, rHash };
}
