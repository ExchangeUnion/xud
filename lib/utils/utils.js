const utils = exports;

utils.copyObject = (source, dest) => {
  const sourceKeys = Object.keys(source);
  for (let n = 0; n < sourceKeys.length; n += 1) {
    const key = sourceKeys[n];
    if (typeof dest[key] === 'object' && !Array.isArray(dest[key])) {
      utils.copyObject(source[key], dest[key]);
    } else {
      dest[key] = source[key]; // eslint-disable-line no-param-reassign
    }
  }
};

utils.nestObject = (obj) => {
  const nestedObject = {};
  const keys = Object.keys(obj);
  for (let n = 0; n < keys.length; n += 1) {
    const key = keys[n];
    const arg = obj[key];
    if (key.indexOf('.') !== -1) {
      const parentKey = key.substring(0, key.indexOf('.'));
      if (!nestedObject[parentKey]) {
        nestedObject[parentKey] = {};
      }
      const childKey = key.substring(key.indexOf('.') + 1);
      nestedObject[parentKey][childKey] = arg;
    } else {
      nestedObject[key] = arg;
    }
  }
  return nestedObject;
};

utils.getTsString = () => (new Date()).toLocaleString();

utils.getPublicMethods = (obj) => {
  const ret = {};
  Object.getOwnPropertyNames(Object.getPrototypeOf(obj)).forEach((name) => {
    const func = obj[name];
    if ((func instanceof Function) && name !== 'constructor' && !name.startsWith('_')) {
      ret[name] = func;
    }
  });
  return ret;
};
