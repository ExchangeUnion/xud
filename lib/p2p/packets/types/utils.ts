import { NodeState } from '../../../types/p2p';
import * as pb from '../../../proto/xudp2p_pb';
import { removeUndefinedProps } from '../../../utils/utils';

export const createPbNodeState = (nodeState: NodeState): pb.NodeState => {
  const pbNodeState = new pb.NodeState();
  pbNodeState.setVersion(nodeState.version);
  pbNodeState.setNodepubkey(nodeState.nodePubKey);
  pbNodeState.setPairsList(nodeState.pairs);
  pbNodeState.setAddressesList(nodeState.addresses!.map((addr) => {
    const pbAddr = new pb.Address();
    pbAddr.setHost(addr.host);
    pbAddr.setPort(addr.port);
    return pbAddr;
  }));
  pbNodeState.setRaidenaddress(nodeState.raidenAddress!);
  pbNodeState.setLndbtcpubkey(nodeState.lndbtcPubKey!);
  pbNodeState.setLndltcpubkey(nodeState.lndltcPubKey!);

  return pbNodeState;
};

export const validatePbNodeStateObj = (obj: pb.NodeState.AsObject | undefined): boolean => {
  return !!(obj
    && obj.nodepubkey
    && obj.pairsList
    && obj.addressesList.filter(addr => addr.host).length === obj.addressesList.length
  );
};

export const convertNodeState = (obj: pb.NodeState.AsObject): NodeState => {
  return removeUndefinedProps({
    version: obj.version,
    nodePubKey: obj.nodepubkey,
    pairs: obj.pairsList,
    addresses: obj.addressesList,
    raidenAddress: obj.raidenaddress || undefined,
    lndbtcPubKey: obj.lndbtcpubkey || undefined,
    lndltcPubKey: obj.lndltcpubkey || undefined,
  });
};
