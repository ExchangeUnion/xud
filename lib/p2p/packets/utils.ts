import * as pb from '../../proto/xudp2p_pb';
import { removeUndefinedProps, convertKvpArrayToKvps, setObjectToMap } from '../../utils/utils';
import { NodeState } from '../types';

export const validateNodeState = (nodeState?: pb.NodeState.AsObject) => {
  // TODO: validate that pairsList does not contain duplicates
  return !!(
    nodeState &&
    nodeState.pairsList &&
    nodeState.lndPubKeysMap &&
    nodeState.tokenIdentifiersMap &&
    nodeState.lndUrisMap &&
    nodeState.addressesList.every((addr) => !!addr.host)
  );
};

export const convertNodeState = (nodeState: pb.NodeState.AsObject) => {
  return removeUndefinedProps({
    pairs: nodeState.pairsList,
    addresses: nodeState.addressesList,
    connextIdentifier: nodeState.connextIdentifier,
    lndPubKeys: convertKvpArrayToKvps(nodeState.lndPubKeysMap),
    lndUris: convertLndUris(nodeState.lndUrisMap),
    tokenIdentifiers: convertKvpArrayToKvps(nodeState.tokenIdentifiersMap),
  });
};

const convertLndUris = (kvpArray: [string, pb.LndUris.AsObject][]): { [key: string]: string[] } => {
  const kvps: { [key: string]: string[] } = {};
  kvpArray.forEach((kvp) => {
    kvps[kvp[0]] = kvp[1].lndUriList;
  });

  return kvps;
};

const setLndUrisMap = (obj: any, map: { set: (key: string, value: any) => any }) => {
  Object.keys(obj).forEach((key) => {
    if (obj[key] !== undefined) {
      const lndUris = new pb.LndUris();
      lndUris.setLndUriList(obj[key]);
      map.set(key, lndUris);
    }
  });
};

export const serializeNodeState = (nodeState: NodeState): pb.NodeState => {
  const pbNodeState = new pb.NodeState();
  pbNodeState.setPairsList(nodeState.pairs);
  pbNodeState.setAddressesList(
    nodeState.addresses.map((addr) => {
      const pbAddr = new pb.Address();
      pbAddr.setHost(addr.host);
      pbAddr.setPort(addr.port);
      return pbAddr;
    }),
  );
  pbNodeState.setConnextIdentifier(nodeState.connextIdentifier);
  if (nodeState.lndPubKeys) {
    setObjectToMap(nodeState.lndPubKeys, pbNodeState.getLndPubKeysMap());
  }
  if (nodeState.lndUris) {
    setLndUrisMap(nodeState.lndUris, pbNodeState.getLndUrisMap());
  }
  if (nodeState.tokenIdentifiers) {
    setObjectToMap(nodeState.tokenIdentifiers, pbNodeState.getTokenIdentifiersMap());
  }
  return pbNodeState;
};
