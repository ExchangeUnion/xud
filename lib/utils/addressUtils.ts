import { Socket } from 'net';
import assert from 'assert';
import { Address } from '../p2p/types';

/** Helper methods for interacting with the [[Address]] type. */
const addressUtils = {
  /**
   * Create an [[Address]] using the remote host and port of a socket.
   */
  fromSocket: (socket: Socket): Address => {
    const { remoteAddress, remotePort } = socket;
    assert(remoteAddress, 'socket must have a remoteAddress value');
    assert(remotePort, 'socket must have a remotePort value');
    return { host: remoteAddress, port: remotePort };
  },

  /**
   * Create an [[Address]] from a string.
   * @param addressString a string in the "{host}:{port}" format
   * @param port a port number to use if no port is specified in the string, defaults to 8885
   */
  fromString: (addressString: string, port = 8885): Address => {
    const arr = addressString.split(':');
    return {
      host: arr[0],
      port: arr[1] ? parseInt(arr[1], 10) : port,
    };
  },

  /** Convert an [[Address]] to a string in the "{host}:{port}" format. */
  toString: (address: Address) => `${address.host}:${address.port}`,

  /** Checks whether two [[Address]] instances are equal, based solely on `host` and `port` fields */
  areEqual: (a: Address, b: Address) => a.host === b.host && a.port === b.port,

  sortByLastConnected: (addresses: Address[]) => {
    return [...addresses].sort((a, b) => {
      if (!a.lastConnected) return 1;
      if (!b.lastConnected) return -1;
      return b.lastConnected - a.lastConnected;
    });
  },
};

export default addressUtils;
