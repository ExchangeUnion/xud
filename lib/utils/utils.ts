import http from 'http';
// @ts-ignore
import createKeccakHash from 'keccak';
import moment from 'moment';
import os from 'os';
import path from 'path';
import { promisify } from 'util';
import { Pair, SortableOrder } from '../orderbook/types';
import p2pErrors from '../p2p/errors';

const MAX_DECIMAL_PLACES = 12;

/**
 * Gets the external IP of the node.
 */
export const getExternalIp = () => {
  return new Promise<string>((resolve, reject) => {
    http
      .get('http://ipv4.icanhazip.com/', (res) => {
        let body = '';

        res.on('data', (chunk) => {
          body += chunk;
        });
        res.on('end', () => {
          // Removes new line at the end of the string
          body = body.trimRight();
          resolve(body);
        });
        res.on('error', (err: Error) => {
          reject(p2pErrors.EXTERNAL_IP_UNRETRIEVABLE(err));
        });
      })
      .on('error', (err: Error) => {
        reject(p2pErrors.EXTERNAL_IP_UNRETRIEVABLE(err));
      })
      .on('error', (err: Error) => {
        reject(p2pErrors.EXTERNAL_IP_UNRETRIEVABLE(err));
      });
  });
};

/**
 * Check whether a variable is a non-array object
 */
export const isObject = (val: any): boolean => {
  return val && typeof val === 'object' && !Array.isArray(val);
};

/**
 * Check whether a variable is an empty object
 */
export const isEmptyObject = (val: any): boolean => {
  return isObject(val) && Object.keys(val).length === 0;
};

/** Get the current date in the given dateFormat, if not provided formats with `YYYY-MM-DD hh:mm:ss.sss`.
 */
export const getTsString = (dateFormat?: string): string => moment().format(dateFormat || 'YYYY-MM-DD hh:mm:ss.sss');

/**
 * Recursively merge properties from different sources into a target object, overriding any
 * existing properties.
 * @param target the destination object to merge into.
 * @param sources the sources objects to copy from.
 */
export const deepMerge = (target: any, ...sources: any[]): object => {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach((key) => {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        deepMerge(target[key], source[key]);
      } else if (source[key] !== undefined) {
        Object.assign(target, { [key]: source[key] });
      }
    });
  }

  return deepMerge(target, ...sources);
};

/**
 * Get all methods from an object whose name doesn't start with an underscore.
 */
export const getPublicMethods = (obj: any): any => {
  const ret: any = {};
  Object.getOwnPropertyNames(Object.getPrototypeOf(obj)).forEach((name) => {
    const func = obj[name];
    if (func instanceof Function && name !== 'constructor' && !name.startsWith('_')) {
      ret[name] = func;
    }
  });
  return ret;
};

export const groupBy = (arr: object[], keyGetter: (item: any) => string | number): any => {
  const ret: any = {};
  arr.forEach((item) => {
    const key = keyGetter(item);
    const group = ret[key];
    if (!group) {
      ret[key] = [item];
    } else {
      group.push(item);
    }
  });
  return ret;
};

/** Returns true if number has more than MAX_DECIMAL_PLACES, false otherwise. */
export const checkDecimalPlaces = (digits: number) => {
  const fixed = Number(digits.toFixed(MAX_DECIMAL_PLACES));
  return fixed < digits;
};

/**
 * Get current time in unix time (milliseconds).
 */
export const ms = (): number => {
  return Date.now();
};

/**
 * Convert a pair's base currency and quote currency to a ticker symbol pair id.
 */
export const derivePairId = (pair: Pair) => {
  return `${pair.baseCurrency}/${pair.quoteCurrency}`;
};

/**
 * A simplified copy of lodash's isPlainObject;
 *
 * A plain object is;
 * - prototype should be [object Object]
 * - shouldn't be null
 * - its type should be 'object' (does extra check because typeof null == object)
 *
 * Examples;
 * isPlainObject(new Foo); => false
 * isPlainObject([1, 2, 3]); => false
 * isPlainObject({ 'x': 0, 'y': 0 }); => true
 * isPlainObject(Object.create(null)); => true
 */
export const isPlainObject = (obj: any) => {
  if (typeof obj !== 'object' || obj === null || Object.prototype.toString.call(obj) !== '[object Object]') {
    return false;
  }
  if (Object.getPrototypeOf(obj) === null) {
    return true;
  }
  let proto = obj;
  while (Object.getPrototypeOf(proto) !== null) {
    proto = Object.getPrototypeOf(proto);
  }
  return Object.getPrototypeOf(obj) === proto;
};

/** A promisified wrapper for the NodeJS `setTimeout` method. */
export const setTimeoutPromise = promisify(setTimeout);

export const removeUndefinedProps = <T>(typedObj: T): T => {
  const obj = typedObj as any;
  Object.keys(obj).forEach((key) => {
    if (obj[key] === undefined) {
      delete obj[key];
    } else if (typeof obj[key] === 'object') {
      removeUndefinedProps(obj[key]);
    }
  });

  return obj;
};

export const setObjectToMap = (obj: any, map: { set: (key: string, value: any) => any }) => {
  Object.keys(obj).forEach((key) => {
    if (obj[key] !== undefined) {
      map.set(key, obj[key]);
    }
  });
};

/**
 * Converts an array of key value pair arrays into an object with the key value pairs.
 */
export const convertKvpArrayToKvps = <T>(kvpArray: [string, T][]): { [key: string]: T } => {
  const kvps: { [key: string]: T } = {};
  kvpArray.forEach((kvp) => {
    kvps[kvp[0]] = kvp[1]; // eslint-disable-line prefer-destructuring
  });

  return kvps;
};

export const sortOrders = <T extends SortableOrder>(orders: T[], isBuy: boolean): T[] => {
  return orders.sort((a: T, b: T) => {
    if (a.price === b.price) {
      return a.createdAt - b.createdAt;
    }
    return isBuy ? a.price - b.price : b.price - a.price;
  });
};

export const base64ToHex = (b64: string) => {
  return Buffer.from(b64, 'base64').toString('hex');
};

export const hexToUint8Array = (hex: string) => {
  return Uint8Array.from(Buffer.from(hex, 'hex'));
};

export const uint8ArrayToHex = (uint8: Uint8Array) => {
  return Buffer.from(uint8).toString('hex');
};

/**
 * Converts input to EIP55 format.
 * Prints the ith digit in uppercase if it's a letter and the 4*ith bit of the hash of the lowercase hexadecimal address is 1
 * otherwise prints it in lowercase.
 * Example: '0xfb6916095ca1df60bb79ce92ce3ea74c37c5d359' is converted to '0xfB6916095ca1df60bB79Ce92cE3Ea74c37c5d359'
 */
export const toEip55Address = (address: string) => {
  const lowercaseAddress = address.toLowerCase().replace('0x', '');
  const hash = createKeccakHash('keccak256').update(lowercaseAddress).digest('hex');
  let ret = '0x';

  for (let i = 0; i < lowercaseAddress.length; i += 1) {
    if (parseInt(hash[i], 16) >= 8) {
      ret += lowercaseAddress[i].toUpperCase();
    } else {
      ret += lowercaseAddress[i];
    }
  }
  return ret;
};

export const getDefaultBackupDir = () => {
  switch (os.platform()) {
    case 'win32':
      return path.join(process.env.LOCALAPPDATA!, 'Xud Backup');
    default:
      return path.join(process.env.HOME!, '.xud-backup');
  }
};

/**
 * A utility function to parse the payload from an http response.
 */
export async function parseResponseBody<T>(res: http.IncomingMessage): Promise<T> {
  res.setEncoding('utf8');
  return new Promise<T>((resolve, reject) => {
    let body = '';
    res.on('data', (chunk) => {
      body += chunk;
    });
    res.on('end', () => {
      resolve(JSON.parse(body));
    });
    res.on('error', (err) => {
      reject(err);
    });
  });
}
