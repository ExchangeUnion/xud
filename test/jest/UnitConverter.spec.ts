import { getUnitConverter } from '../utils';

export const UNITS_PER_CURRENCY: { [key: string]: bigint } = {
  BTC: 1n,
  LTC: 1n,
  ETH: 10n ** 10n,
};

describe('UnitConverter', () => {
  const unitConverter = getUnitConverter();

  describe('amountToUnits', () => {
    test('converts BTC amount to units', () => {
      const amount = 99999999;
      expect(
        unitConverter.amountToUnits({
          amount,
          currency: 'BTC',
        }),
      ).toEqual(BigInt(amount));
    });

    test('converts ETH amount to units', () => {
      expect(
        unitConverter.amountToUnits({
          amount: 7500000,
          currency: 'ETH',
        }),
      ).toEqual(75000000000000000n);
    });

    test('throws error upon unknown currency', () => {
      expect.assertions(1);
      try {
        unitConverter.amountToUnits({
          amount: 123,
          currency: 'ABC',
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('unitsToAmount', () => {
    test('converts BTC units to amount', () => {
      const units = 99999999n;
      expect(
        unitConverter.unitsToAmount({
          units,
          currency: 'BTC',
        }),
      ).toEqual(Number(units));
    });

    test('converts ETH units to amount', () => {
      expect(
        unitConverter.unitsToAmount({
          units: 75000000000000000n,
          currency: 'ETH',
        }),
      ).toEqual(7500000);
    });

    test('throws error upon unknown currency', () => {
      expect.assertions(1);
      try {
        unitConverter.unitsToAmount({
          units: 123n,
          currency: 'ABC',
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('calculateInboundOutboundAmounts', () => {
    const pairId = 'LTC/BTC';
    const quantity = 250000;
    const bigQuantity = 250000n;
    const price = 0.01;
    const priceDivisor = 100n;

    test('calculate inbound and outbound amounts and currencies for a buy order', () => {
      const {
        inboundCurrency,
        inboundAmount,
        outboundCurrency,
        outboundAmount,
        inboundUnits,
        outboundUnits,
      } = unitConverter.calculateInboundOutboundAmounts(quantity, price, true, pairId);
      expect(inboundCurrency).toEqual('LTC');
      expect(inboundAmount).toEqual(quantity);
      expect(inboundUnits).toEqual(UNITS_PER_CURRENCY['LTC'] * bigQuantity);
      expect(outboundCurrency).toEqual('BTC');
      expect(outboundAmount).toEqual(quantity * price);
      expect(outboundUnits).toEqual((UNITS_PER_CURRENCY['BTC'] * bigQuantity) / priceDivisor);
    });

    test('calculate inbound and outbound amounts and currencies for a sell order', () => {
      const {
        inboundCurrency,
        inboundAmount,
        outboundCurrency,
        outboundAmount,
        inboundUnits,
        outboundUnits,
      } = unitConverter.calculateInboundOutboundAmounts(quantity, price, false, pairId);
      expect(inboundCurrency).toEqual('BTC');
      expect(inboundAmount).toEqual(quantity * price);
      expect(inboundUnits).toEqual((UNITS_PER_CURRENCY['BTC'] * bigQuantity) / priceDivisor);
      expect(outboundCurrency).toEqual('LTC');
      expect(outboundAmount).toEqual(quantity);
      expect(outboundUnits).toEqual(UNITS_PER_CURRENCY['LTC'] * bigQuantity);
    });

    test('calculate 0 outbound amount for a market buy order', () => {
      const { outboundCurrency, outboundAmount, outboundUnits } = unitConverter.calculateInboundOutboundAmounts(
        quantity,
        0,
        true,
        pairId,
      );
      expect(outboundCurrency).toEqual('BTC');
      expect(outboundAmount).toEqual(0);
      expect(outboundUnits).toEqual(0n);
    });

    test('calculate 0 inbound amount for a market sell order', () => {
      const { inboundCurrency, inboundAmount, inboundUnits } = unitConverter.calculateInboundOutboundAmounts(
        quantity,
        Number.POSITIVE_INFINITY,
        false,
        pairId,
      );
      expect(inboundCurrency).toEqual('BTC');
      expect(inboundAmount).toEqual(0);
      expect(inboundUnits).toEqual(0n);
    });

    test('calculate inbound and outbound amounts and currencies for a Connext order', () => {
      const {
        inboundCurrency,
        inboundAmount,
        outboundCurrency,
        outboundAmount,
        inboundUnits,
        outboundUnits,
      } = unitConverter.calculateInboundOutboundAmounts(quantity, price, true, 'ETH/BTC');
      expect(inboundCurrency).toEqual('ETH');
      expect(inboundAmount).toEqual(quantity);
      expect(inboundUnits).toEqual(UNITS_PER_CURRENCY['ETH'] * bigQuantity);
      expect(outboundCurrency).toEqual('BTC');
      expect(outboundAmount).toEqual(quantity * price);
      expect(outboundUnits).toEqual((UNITS_PER_CURRENCY['BTC'] * bigQuantity) / priceDivisor);
    });
  });
});
