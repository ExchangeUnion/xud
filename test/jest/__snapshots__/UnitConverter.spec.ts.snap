// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitConverter amountToUnits throws error upon unknown currency 1`] = `[Error: cannot convert ABC amount of 123 to units because decimal places per currency was not found in the database]`;

exports[`UnitConverter unitsToAmount throws error upon unknown currency 1`] = `[Error: cannot convert ABC units of 123 to units because decimal places per currency was not found in the database]`;
